### PR TITLE
chore(vrl): RFC for iteration support

### LINKED ISSUE #8381: VRL iteration support

### PR DESCRIPTION
**Pull request by JeanMertz**
closes #8381

👀 [**RENDERED**](https://github.com/vectordotdev/vector/blob/f1b3c52d4986ad4c32a06213e342ffbde6830e1e/rfcs/2021-08-29-8381-vrl-iteration-support.md)

### CODE REVIEW

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (StephenWakely) [inline rfcs/2021-08-29-8381-vrl-iteration-support.md:419]: A minor detail, but I don't think this is quite right. Scoping is more about where the variables are declared and available rather than what happens to the previous values.

The confusion is that in VRL we don't have the concept of declaring variables, so we bring them into scope the first type a value is set.

I would argue that in the previous example, the `foo = "bar"` statement declares the variable, then when we enter the block that definition is moved into the scope of the block and `foo = "baz"` just sets the value of the original `foo`. So with or without lexical scoping in that example `foo` should return "baz".

Where it would break would be if you had:

```coffee
{
  foo = "baz"
}
foo
```

In this case, `foo` would not exist outside the block, currently it would and would return "baz".

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (StephenWakely) [inline rfcs/2021-08-29-8381-vrl-iteration-support.md:445]: I would prefer, 

```
value1 = 1.1
value2 = undefined
value3 = undefined
```

As that is what I would expect from most programming languages. The var hoisting you get in javascript which is what you propose here with `value3` was pretty unpopular - for good reason.

I also think it would be more straighforward to implement as we just implement our variable list as a stack and consistent rules should appy when accessing the variables. Variables are accessed by checking for existence in the top level of the stack, if they don't exist there then check in the previous layer. If there is no previous layer, then error. When exiting the block we wouldn't need to concern ourselves with any copying values and types between layers of the stack.

The concern with the block parameter variables can be addressed by treating the parameters as new definitions, and thus new variables. Essentially, the block just pushes new definitions for these variable names to the scope stack, so accessing the variables types wouldn't fall through to the previous level.

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (StephenWakely) [inline rfcs/2021-08-29-8381-vrl-iteration-support.md:493]: I slightly prefer being more explicit and returning an object rather than an array here:

```coffee
{ key = key, value = value }
```

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (StephenWakely) [inline rfcs/2021-08-29-8381-vrl-iteration-support.md:537]: I think we should require the blocks to be surrounded by `{...}`. I believe Ruby requires you to put the `{` before the parameter list which should fix this.

```ruby
fn(a) { |b| c }
```

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (StephenWakely) [inline rfcs/2021-08-29-8381-vrl-iteration-support.md:556]: https://www.tutorialspoint.com/ruby/ruby_blocks.htm

**Review by StephenWakely — COMMENTED** [review]
Very excited to see this land. I think this is very much on the right track!

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (binarylogic) [inline rfcs/2021-08-29-8381-vrl-iteration-support.md:110]: 👍 I love this to upload the safety principle.

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (binarylogic) [reply rfcs/2021-08-29-8381-vrl-iteration-support.md:445]: I agree with this.

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (binarylogic) [inline rfcs/2021-08-29-8381-vrl-iteration-support.md:552]: Just noting that all of these drawbacks are worth the cost to unblock the use cases iteration is solving.

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (binarylogic) [inline rfcs/2021-08-29-8381-vrl-iteration-support.md:647]: I'm ok with an array as long as the compiler strictly checks this. And I agree, sticking with JSON types is important, otherwise we introduce other issues with mapping tuples to JSON (as you mentioned). I also like @StephenWakely's suggestion of a map as well.

**Review by binarylogic — APPROVED** [review]
This is an excellent RFC, nice work. Giving my approval on the general proposal of closures. I'd like to see all of the comments resolved and the TODOs complete.

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (spencergilbert) [reply rfcs/2021-08-29-8381-vrl-iteration-support.md:146]: Presumably when https://github.com/timberio/vector/pull/8912 is merged 🤔

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (spencergilbert) [reply rfcs/2021-08-29-8381-vrl-iteration-support.md:445]: Same

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (spencergilbert) [reply rfcs/2021-08-29-8381-vrl-iteration-support.md:493]: Being more explicit aligns with the VRL design doc - perhaps treated as function arguments today?

```coffee
{ key = key, value = value }
{ key, value }
{ value = value, key = key }
```

Would all be equivalent.

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (spencergilbert) [inline rfcs/2021-08-29-8381-vrl-iteration-support.md:520]: 👍

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (spencergilbert) [inline rfcs/2021-08-29-8381-vrl-iteration-support.md:635]: Reminds me of the pain in Helm templating where you use `$` instead for "local" things, which I don't love.

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (JeanMertz) [reply rfcs/2021-08-29-8381-vrl-iteration-support.md:159]: It will 👍 I'll add a note.

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (JeanMertz) [reply rfcs/2021-08-29-8381-vrl-iteration-support.md:419]: Agreed. I do think without variable assignment syntax (e.g. let-bindings), you can't do certain things you might want to do in terms of assigning/re-assigning, but I don't think we _need_ to introduce that in this RFC, if at all.

I'll update the RFC to match what you noted here.

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (JeanMertz) [reply rfcs/2021-08-29-8381-vrl-iteration-support.md:423]: Agreed. I was a fan initially as well, but worried that this RFC would introduce too many new concepts/breaking changes, and since it wasn't _technically_ required, I removed it in a later version of the RFC.

I'll add it back as something we prefer not to punt on, and solve with this RFC.

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (JeanMertz) [reply rfcs/2021-08-29-8381-vrl-iteration-support.md:445]: Agreed, I'll make the change.

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (JeanMertz) [reply rfcs/2021-08-29-8381-vrl-iteration-support.md:493]: I'm torn on this. I somewhat think the required object is _too_ verbose. The second example that @spencergilbert mentions is something @StephenWakely and I discussed offline as well, but I'm not convinced we can make that work, due to parsing ambiguity (`{ foo }` can be both a block expression, or a keyless-object expression).

Rust has the same issue with `(foo)` being either a grouped-expression or a tuple-expression. They solve this by requiring you to add an extra `,` for single-element tuple expressions: `(foo,)`. We could do something similar.

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (JeanMertz) [reply rfcs/2021-08-29-8381-vrl-iteration-support.md:537]: I'm irrationally against that syntax, it "doesn't feel right to me". I agree though that it would help with the ambiguity, so it's worth considering.

I do very much like the non-block closure syntax for simple iterations, but I wouldn't be against not allowing that, at least not in a V1 for simplicity reasons.

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (JeanMertz) [reply rfcs/2021-08-29-8381-vrl-iteration-support.md:647]: - Strictly checking will be there, I'll make a note about it
- Mapping tuples to JSON isn't that strange ([Python can do it](https://pythonexamples.org/run.php?pgm=import+json%0A%0A%23python+tuple%0Amytuple+%3D+%28%22python%22%2C+%22json%22%2C+22%2C+23.04%29%0A%0A%23convert+to+JSON+string%0AjsonStr+%3D+json.dumps%28mytuple%29%0A%0A%23print+json+string%0Aprint%28jsonStr%29%0A%0A%23parse+json%0AjsonArr+%3D+json.loads%28jsonStr%29%0A%0A%23print+third+item+of+the+json+array%0Aprint%28jsonArr%5B2%5D%29)) because JSON allows array elements to be of different types, so in effect, a JSON array is similar to a tuple. Obviously, when deserializing a JSON array to a VRL value, you'd have to choose between a tuple or an array. Having said all that, tuples aren't needed, but they also _can_ be implemented cleanly.
- Using objects is definitely a possibility, it's a bit more verbose (and perhaps _too_ verbose), but it is also more clear, and better fits the rules we list in our design document.

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (pablosichert) [reply rfcs/2021-08-29-8381-vrl-iteration-support.md:423]: I can see the concern of not wanting to overload this RFC, but would also very be in favor of having well-defined lexical scoping.

Would it be possible we first do an RFC for lexical scoping, and base this RFC on it? The initial RFC wouldn't need to mention closures yet.

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (pablosichert) [reply rfcs/2021-08-29-8381-vrl-iteration-support.md:445]: Adding to https://github.com/timberio/vector/pull/9001/files#r701814426, I'd also prefer `value1 = undefined` if we can make it.

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (pablosichert) [inline rfcs/2021-08-29-8381-vrl-iteration-support.md:734]: Regarding the discussion in https://github.com/timberio/vector/pull/9001/files#r700935702, wouldn't this case here be impossible to disambiguate when parsing?

Specifically, if `|>` relates to `is_ip(ip)` or `filter(.ip) |_, ip| is_ip(ip)`

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (spencergilbert) [reply rfcs/2021-08-29-8381-vrl-iteration-support.md:493]: That's fair - it's not a hill I'm interested to die on 😄 

```coffee
[key, value]
[value, key]
```

Isn't  all that different 😉

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (JeanMertz) [reply rfcs/2021-08-29-8381-vrl-iteration-support.md:734]: It wouldn't. It would require defining operator precedence, but we already do that.

That is, the pipeline operator would either be part of the inner expression, or it wouldn't. In the above example, it _wouldn't_.

Having said that, I agree it's at least ambiguous to readers, so that would be an argument against allowing this short-hand form of writing closures.

I think it's still nice to do (for example):

```coffee
. = map(.) |k, v| [upcase(k), v]
```

But it's not _terrible_ if that weren't possible:

```coffee
. = map(.) |k, v| { [upcase(k), v] }
```

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (spencergilbert) [reply rfcs/2021-08-29-8381-vrl-iteration-support.md:423]: @pablosichert's suggestion seems reasonable - especially given we don't _seem_ rushed to land an iteration solution yet, just planning the implementation (I don't believe the work is scheduled in the near term?)

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (spencergilbert) [reply rfcs/2021-08-29-8381-vrl-iteration-support.md:734]: That's a good catch - does seem like we'll need some sort of identifier here `->` or `{}`

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (pablosichert) [inline rfcs/2021-08-29-8381-vrl-iteration-support.md:364]: On the other hand, if we'd allow closures in a variable any special handling for recursion (the `recursive: true` argument or `recursive for` construct) would not be needed at the language level.

```
upcase_keys = |key, value| {
  [upcase(key), if is_object(value) {
    upcase_keys(value)
  } else {
    value
  }]
}

map(., upcase_keys)
```

I briefly talked about it with @JeanMertz and he made the case that the `recursive` constructs in this RFC might be more easily understood by users that are not very familiar with programming languages - which I agree with.

Interested in what the others think about this.

Comment (JeanMertz) [reply rfcs/2021-08-29-8381-vrl-iteration-support.md:364]: The argument you attributed to me is correct, and harks back to  the "**[design the feature for the intended target audience](https://github.com/timberio/vector/blob/jean/vrl-design-doc/lib/vrl/DESIGN.md#target-audience)**" section of the VRL design document.

There are more (more pressing) reasons why I'm against allowing this, though.

The most important one, is that allowing this can result in poor-performing code, even resulting in infinite-loops. The most important rule in the VRL design document states **[Safety and performance over ease of use](https://github.com/timberio/vector/blob/jean/vrl-design-doc/lib/vrl/DESIGN.md#the-zen-of-vrl)**. Which this technique would violate.

Secondly, I want to keep variable and target assignment (e.g. `foo = true` and `.foo = true`) semantically equal in VRL. You can assign any value to a variable that you can also assign to an event field, and vice-versa.

When you assign a closure to a variable, you are assigning _code that is optionally evaluated at runtime_. We cannot do this for event field assignment without seeing a significant increase in cognitive complexity on assignment rules, or by rejecting assigning closures to event fields, which again makes this a non-starter to me.

Comment (JeanMertz) [reply rfcs/2021-08-29-8381-vrl-iteration-support.md:364]: I'll add the above comments to the RFC, to better explain why this was rejected as a viable option.

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (JeanMertz) [reply rfcs/2021-08-29-8381-vrl-iteration-support.md]: Having now added lexical scoping to this RFC in 13ab8eaf45b211991c4aad87e9cbc556a2126457, I don't _think_ it needs its own RFC (the rules are simple), but I wouldn't mind doing the work (after my holiday) of splitting up this RFC into two RFCs if people would prefer that.

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (JeanMertz) [reply rfcs/2021-08-29-8381-vrl-iteration-support.md]: I came around to:

- not minding us losing short-hand notation of closures, it's only two less characters (and, as @pablosichert mentioned, it can result in more cognitive complexity when it comes to operator precedence).
- not being a fan of the `-> { ... }` syntax

The new syntax is now as you proposed:

```coffee
fn(a) { |b| c }
```

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (JeanMertz) [reply rfcs/2021-08-29-8381-vrl-iteration-support.md:627]: Agreed. Added in 13ab8eaf45b211991c4aad87e9cbc556a2126457.

**Review by spencergilbert — APPROVED** [review]

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (JeanMertz) [reply rfcs/2021-08-29-8381-vrl-iteration-support.md]: Fixed in 84c128bcb

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (JeanMertz) [reply rfcs/2021-08-29-8381-vrl-iteration-support.md]: Yep. Added a note in e80c76b15.

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (JeanMertz) [reply rfcs/2021-08-29-8381-vrl-iteration-support.md]: Added note in da1b44a71bea17e2c883960565a1aafa8f7a7fc1.

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (JeanMertz) [reply rfcs/2021-08-29-8381-vrl-iteration-support.md:628]: :nod: yeah.

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (JeanMertz) [reply rfcs/2021-08-29-8381-vrl-iteration-support.md:731]: Resolved by changing the function-closure syntax, which also removes the short-hand notation. See 1ad39083701c4c6739715490435d761e3a1eecb5.

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (JeanMertz) [reply rfcs/2021-08-29-8381-vrl-iteration-support.md:499]: As noted in Slack, I'm leaning towards moving forward as-is, given that returning an array or object is based on the `map` function implementation, and doesn't need any compiler changes. Additionally, actually building that map function will be (1) the last piece of the implementation puzzle, and (2) relatively easy compared to all the other work, so we don't have to sweat this detail as much as others before we start the implementation work.

**Review by StephenWakely — APPROVED** [review]
Nice work!

I agree, it's not worth holding this up over the decision to return an object or an array as that is down to the function implementation. We could even make it so it accepts either if we really wanted.

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (jszwedko) [reply rfcs/2021-08-29-8381-vrl-iteration-support.md:503]: Yeah, I agree, I think returning an object is a bit verbose. It would be nice to have real tuples, but I think the two-element array is next best.

**Review by jszwedko — APPROVED** [review]
Kudos on this thorough RFC!

I could see users wanting to be able to reuse blocks in the future for complex mapping operations, but I agree with you that it would introduce the risk of infinite recursion. What you have here seems like it should cover the vast majority of use-cases.

**Comment by dbcfd** [comment]
I posted some comments in https://github.com/vectordotdev/vector/issues/6031#issuecomment-919525821

Given the similarities to rust that already exist in VRL, and the simplicity of understanding `for x in coll` I prefer it to the `map` approach. This is extensible to objects as well through the introduction of a tuple return without destructuring sugar.

I think implementing something simple like this would then allow for iteration, as additional use cases are presented.

**Comment by JeanMertz** [comment]
> Given the similarities to rust that already exist in VRL, and the simplicity of understanding `for x in coll` I prefer it to the `map` approach.

Thank you for the input @dbcfd. I agree that the familiarity and simplicity of for-loops is a big plus. Unfortunately, as [described in the RFC](https://github.com/vectordotdev/vector/blob/540eddf911487cfa6c74811b1df77038a739afd7/rfcs/2021-08-29-8381-vrl-iteration-support.md#for-loop), there are downsides to this solution specific to VRL that mostly erase both those aspects of a simple for-loop, which is why we rejected that solution in the end.

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (JeanMertz) [reply rfcs/2021-08-29-8381-vrl-iteration-support.md:996]: Agreed. I wanted to keep it out of this RFC, but expect to either have a mini-RFC ready when we start implementing this one to add this feature, or implement it straight-away shortly after landing the first iteration.

File: rfcs/2021-08-29-8381-vrl-iteration-support.md
Comment (JeanMertz) [reply rfcs/2021-08-29-8381-vrl-iteration-support.md:265]: Correct.

**Pull request merged**
